src/System.Memory/src/System/Buffers/ReadOnlyBuffer.cs

```diff
@@ -0,0 +1,377 @@
+        /// <param name="length">The length of the slice</param>
+        public ReadOnlyBuffer<T> Slice(SequencePosition start, int length)
+        {
+            BoundsCheck(BufferEnd, start);
+
+            var end = Seek(start, BufferEnd, length, false);
+            return SliceImpl(start, end);
+        }
+
+        /// <summary>
+        /// Forms a slice out of the given <see cref="ReadOnlyBuffer{T}"/>, beginning at <paramref name="start"/>, ending at <paramref name="end"/> (inclusive).
+        /// </summary>
+        /// <param name="start">The starting (inclusive) <see cref="SequencePosition"/> at which to begin this slice.</param>
+        /// <param name="end">The ending (inclusive) <see cref="SequencePosition"/> of the slice</param>
+        public ReadOnlyBuffer<T> Slice(SequencePosition start, SequencePosition end)
+        {
+            BoundsCheck(BufferEnd, end);
+            BoundsCheck(end, start);
+
+            return SliceImpl(start, end);
+        }
+
+        /// <summary>
+        /// Forms a slice out of the given <see cref="ReadOnlyBuffer{T}"/>, beginning at <paramref name="start"/>, ending at the existing <see cref="ReadOnlyBuffer{T}"/>'s end.
+        /// </summary>
+        /// <param name="start">The starting (inclusive) <see cref="SequencePosition"/> at which to begin this slice.</param>
+        public ReadOnlyBuffer<T> Slice(SequencePosition start)
+        {
+            BoundsCheck(BufferEnd, start);
+
+            return SliceImpl(start, BufferEnd);
+        }
+
+        /// <summary>
+        /// Forms a slice out of the given <see cref="ReadOnlyBuffer{T}"/>, beginning at <paramref name="offset"/>, ending at the existing <see cref="ReadOnlyBuffer{T}"/>'s end.
+        /// </summary>
+        /// <param name="offset">The start index at which to begin this slice.</param>
+        public ReadOnlyBuffer<T> Slice(long offset)
+        {
+            if (offset == 0) return this;
+
+            var begin = Seek(BufferStart, BufferEnd, offset, false);
+            return SliceImpl(begin, BufferEnd);
+        }
+
+        public override string ToString()
+        {
+            if (typeof(T) == typeof(byte))
+            {
+                if (this is ReadOnlyBuffer<byte> bytes)
+                {
+                    var sb = new StringBuilder();
+                    foreach (var buffer in bytes)
+                    {
+                        SpanLiteralExtensions.AppendAsLiteral(buffer.Span, sb);
+                    }
+                    return sb.ToString();
+                }
+            }
+            return base.ToString();
+        }
+
+        /// <summary>
+        /// Returns an enumerator over the <see cref="ReadOnlyBuffer{T}"/>
+        /// </summary>
+        public Enumerator GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
+
+        public SequencePosition GetPosition(SequencePosition origin, long offset)
```

Comment:
What if offset is too large? Add checks here and elsewhere.

Reply:
Seek would throw if offset it too large